crypto/threads_pthread.c

```diff
@@ -42,12 +42,401 @@
+ * a given lock
+ * users is the 64 bit value that stores
+ * the READERS/ID as defined
+ * above
+ *
+ */
+struct rcu_qp {
+    uint64_t users;
+};
+
+/*
+ * This is the per thread tracking data
+ * that is assigned to each thread participating
+ * in an rcu qp
+ *
+ * qp points to the qp that it last acquired
+ *
+ * count is the threads current recursion count
+ */
+struct rcu_thr_data {
+    struct rcu_qp *qp;
+    int count;
+};
+
+/*
+ * This is the internal version of a CRYPTO_RCU_LOCK
+ * it is cast from CRYPTO_RCU_LOCK
+ */
+struct rcu_lock_internal {
+    CRYPTO_RCU_LOCK lock; /* Must be first */
+    struct rcu_cb_item *cb_items;
+    uint32_t id_ctr;
+    struct rcu_qp *qp_group;
+    size_t group_count;
+    uint32_t reader_idx;
+    uint32_t next_to_retire;
+    uint32_t current_alloc_idx;
+    uint32_t writers_alloced;
+    pthread_mutex_t write_lock;
+    pthread_mutex_t alloc_lock;
+    pthread_cond_t alloc_signal;
+    pthread_mutex_t prior_lock;
+    pthread_cond_t prior_signal;
+};
+
+/*
+ * Called on thread exit to free the pthread key
+ * associated with this thread, if any
+ */
+static void free_rcu_thr_data(void *ptr)
+{
+    struct rcu_thr_data *data = ptr;
+
+    assert(data != NULL);
+    CRYPTO_free(data, __FILE__, __LINE__);
+}
+
+/*
+ * Called from OPENSSL_SSL_init to initalize the
+ * rcu lock system
+ */
+void CRYPTO_THREAD_rcu_init(void)
+{
+    pthread_key_create(&rcu_thr_key, free_rcu_thr_data);
+}
+
+/*
+ * Read side acquisition of the current qp
+ */
+static struct rcu_qp *get_hold_current_qp(CRYPTO_RCU_LOCK *lock)
+{
+    uint32_t qp_idx;
+    struct rcu_lock_internal *ilock =
+        (struct rcu_lock_internal *)lock;
+
+    /*
+     * get the current qp index
+     */
+try_again:
+    qp_idx = __atomic_load_n(&ilock->reader_idx, __ATOMIC_SEQ_CST);
+    __atomic_add_fetch(&ilock->qp_group[qp_idx].users, VAL_READER,
+                       __ATOMIC_SEQ_CST);
+    if (qp_idx != __atomic_load_n(&ilock->reader_idx, __ATOMIC_SEQ_CST)) {
+        __atomic_sub_fetch(&ilock->qp_group[qp_idx].users, VAL_READER,
+                       __ATOMIC_SEQ_CST);
+        goto try_again;
+    }
+
+    return (struct rcu_qp *)&ilock->qp_group[qp_idx];
+}
+
+/*
+ * Public READ lock api
+ */
+void CRYPTO_THREAD_rcu_read_lock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_thr_data *data;
+
+    /*
+     * we're going to access current_qp here so ask the
+     * processor to fetch it
+     */
+    data = pthread_getspecific(rcu_thr_key);
+
+    if (data == NULL) {
+        data = OPENSSL_zalloc(sizeof(struct rcu_thr_data));
+        assert(data != NULL);
+        pthread_setspecific(rcu_thr_key, data);
+    }
+
+    if (data->qp == NULL)
+        data->qp = get_hold_current_qp(lock);
+
+    /* inc our local count */
+    data->count++;
+
+    /* check for underflow condition */
+    assert(data->count >= 0);
+}
+
+/*
+ * Public READ unlock api
+ */
+void CRYPTO_THREAD_rcu_read_unlock(CRYPTO_RCU_LOCK *lock)
```

Comment:
Why does this implementation make no reference to the lock that we are unlocking??